code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceVolumeLookupPreCreate.java

```diff
@@ -0,0 +1,111 @@
+import io.cattle.platform.core.model.Instance;
+import io.cattle.platform.core.model.StoragePool;
+import io.cattle.platform.core.model.Volume;
+import io.cattle.platform.engine.handler.HandlerResult;
+import io.cattle.platform.engine.handler.ProcessPreListener;
+import io.cattle.platform.engine.process.ProcessInstance;
+import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.object.ObjectManager;
+import io.cattle.platform.object.process.ObjectProcessManager;
+import io.cattle.platform.object.util.DataAccessor;
+import io.cattle.platform.process.common.handler.AbstractObjectProcessLogic;
+import io.cattle.platform.util.exception.ExecutionException;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Named
+public class InstanceVolumeLookupPreCreate extends AbstractObjectProcessLogic implements ProcessPreListener {
+
+    private static final Logger log = LoggerFactory.getLogger(InstanceVolumeLookupPreCreate.class);
+    private static final String LOCAL = "local";
+
+    @Inject
+    ObjectManager objectManager;
+    @Inject
+    ObjectProcessManager processManager;
+    @Inject
+    StoragePoolDao storagePoolDao;
+    @Inject
+    VolumeDao volumeDao;
+
+    @Override
+    public String[] getProcessNames() {
+        return new String[] { "instance.create" };
+    }
+
+    /**
+     * Looks for shared volumes in the dataVolumes field and converts them to dataVolumeMounts.
+     * If volumeDriver is set:
+     *  if set to local, do nothing
+     *  if set to anything else, restrict volume lookup to a matching storage pool
+     *  if blank, do not restrict volume lookup to a particular storage pool.
+     * If volume is found, remove from dataVolumes list and put it in dataVolumeMounts. 
+     * If volume is not found, just leave it in dataVolumes.
+     */
+    @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        Instance instance = (Instance)state.getResource();
+        if (!InstanceConstants.KIND_CONTAINER.equals(instance.getKind())) {
+            return null;
+        }
+
+        String vd = DataAccessor.fieldString(instance, InstanceConstants.FIELD_VOLUME_DRIVER);
+        if (LOCAL.equals(vd)) {
+            return null;
+        }
+
+        Long storagePoolId = null;
+        long accountId = instance.getAccountId();
+        if (StringUtils.isNotEmpty(vd)) {
+            StoragePool sp = storagePoolDao.findStoragePoolByDriverName(accountId, vd);
+            if (sp != null) {
+                storagePoolId = sp.getId();
+            }
+        }
+
+        Map<Object, Object> data = new HashMap<>();
+        List<String> dataVolumes = DataAccessor.fieldStringList(instance, InstanceConstants.FIELD_DATA_VOLUMES);
```

Comment:
Should we add .withDefault(Collections.EMPTY_LIST) here? or may be its better to check for not null when do data.put on lines 85 and 86 to avoid storing null values unnecessarily.